Rewrite patcher in Bash

Miepee opened the pull request:
This PR rewrites the current patcher from Python into Bash.
It has a few advantages over the current Python approach:
- It is more condensed, and more readable
- Since Bash runs on basically every system nowadays (almost always preinstalled too), it is much more portable to other OS should they need a modification of the patcher.
- It has a *fully automatic/headless* CLI approach now, which makes it possible to use it i.e. in CI systems and makes it more similar to other existing unix tools.
- Has a toggle whether to generate an AppImage or not, since some distros / people prefer to not have one.
- Respects a  "PREFIX" like other build tools, which allows the patcher to build the patch into specific directories instead of being hardcoded to "script-dir"

Notes:
- This does not affect the AM2RLauncher, as the launcher does not interact with these tools.
- Should the script be run without arguments, it'll launch the interactive mode which acts basically like the python script acted before.
- I tested AppImage creation and normal creation, with interactive mode and without, as well as with default and custom Prefixes. All of those on my machine at least (feel free to retest though) acted and launched as they should. I tested creation of APKs, it runs fine and plays music as well.
Additionally fixes #28 

Some food for thought from me:
- ~~I need to edit the readme, since it still mentions the python script~~ Done.
- ~~Before attempting to patch to Android, it should be possible to check whether Java is installed in the first place, and error out early with an "Java must be installed in order to patch for Android", instead of going with the current approach of just pretending it's there. Same for Linux(+Android) with xdelta.~~ Done.
- There are currently no sanity checks done for the 1.1 copy. We're just trusting that it's the right one, and it'll just fail on xdelta should it be the wrong one.

---
Asking for review of @Lojemiru as well as @ivarwithoutbones who wanted a headless way of patching for nix packages.

Miepee committed:
Add shell patcher

Miepee committed:
Readme changes, exclude apk from git, slight android cleanage

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+		RESOURCES="${OUTPUT}/assets"
+	fi
+	# Cleanup in case the dirs/files exists
+	cleanupDirs=("$OUTPUT" "$SCRIPT_DIR/AM2R.AppDir" "$SCRIPT_DIR/AM2R-x86_64.AppImage" "$SCRIPT_DIR/AM2RWrapper/" "$SCRIPT_DIR/utilities/android/assets/" "$SCRIPT_DIR/AM2RWrapper.apk")
+	rm -rf "${cleanupDirs[@]}"
```
This doesn't have to be an array as it's not used later. You can just pass all files to `rm`:
```suggestion
	rm -rf "$OUTPUT" "$SCRIPT_DIR/AM2R.AppDir" "$SCRIPT_DIR/AM2R-x86_64.AppImage" "$SCRIPT_DIR/AM2RWrapper/" "$SCRIPT_DIR/utilities/android/assets/" "$SCRIPT_DIR/AM2RWrapper.apk"
```
The comment should already explain what its doing.

Miepee replied:
Was thinking at first that this'd be less readable, then I remembered we do the exact same stuff multiple times later.
So applied locally it is.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+	cleanupDirs=("$OUTPUT" "$SCRIPT_DIR/AM2R.AppDir" "$SCRIPT_DIR/AM2R-x86_64.AppImage" "$SCRIPT_DIR/AM2RWrapper/" "$SCRIPT_DIR/utilities/android/assets/" "$SCRIPT_DIR/AM2RWrapper.apk")
+	rm -rf "${cleanupDirs[@]}"
+
+	mkdir -p "$OUTPUT"
+	mkdir -p "$RESOURCES"
```
```suggestion
	mkdir -p "$OUTPUT" "$RESOURCES"
```

Miepee replied:
Applied locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+	# Check for AM2R_11
+	if [[ -f "$AM2RZIP" ]]; then
+		echo "AM2R_11.zip found!"
+		unzip -q "$AM2RZIP" -d "$OUTPUT"
```
Maybe `OUTPUT` is a bit of a misleading name here, seeing how the final output the script produces doesn't get pushed to it. You could create a temporary directory named `TMPDIR` or something instead with `mktemp -d`, and trap its deletion with `trap "rm -rf $TMPDIR" EXIT`.

Miepee replied:
> seeing how the final output the script produces doesn't get pushed to it.

Hm, from what I can see, it does partially.
For linux local, the resulting stuff will be in $OUTPUT (i.e. OUTPUT/runner, OUTPUT/assets, OUTPUT/.desktop etc.)
For Linux systemwide, the resulting stuff will be in $OUTPUT (i.e. $OUTPUT/am2r.appimage, OUTPUT/icon.png etc.)

Android as far as i can see is the only path that places its stuff into SCRIPT_DIR instead of OUTPUT.

IvarWithoutBones replied:
For consistency I think it'd be best to use the `PREFIX` variable in those cases as well. `PREFIX` should always contain the path to the final output this script produces, I think that would make the script a lot more readable.

IvarWithoutBones replied:
Since the contents of the zip are always copied to create a clean mutable temporary directory, i do think `mktemp -d` makes sense here.

Miepee replied:
Hm, not that sure about that. Currently OUTPUT exists inside PREFIX.
I.e., for a systemwide installation, you have something akin to this:
```
/
|-bin/
  |-...
|-opt/
  |-...
|-share/
  |-...
```
Where `/` is $PREFIX, and `/opt` is $OUTPUT
For a local installation, it'd be more like
```
~/patcher/
|-am2r_15-5
   |-...
```
where `~/patcher/` is $PREFIX and `am2r_15-5` is $OUTPUT. 

So $PREFIX already *is* the final output. Maybe it'd make sense to rename OUTPUT to something else? Since PREFIX is the place where all relevant game files with reside, with OUTPUT actually being more the place for the game data files. Something like `$GAMEDIR` perhaps?

IvarWithoutBones replied:
`GAMEDIR` sounds good yeah, or `ASSETS` maybe?

Miepee replied:
I'll rename it to GAMEDIR, as ASSETS is kinda misleading with the RESOURCES variable (which does point to "$output/assets)"

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+		unzip -q "$AM2RZIP" -d "$OUTPUT"
+	elif [[ -d "$AM2RZIP" ]]; then
+		echo "AM2R_11 folder found!"
+		cp -R "$AM2RZIP" "$OUTPUT"
```
Does this need to get copied? Cant we just set `OUTPUT` to `AM2RZIP`?

Miepee replied:
We do not want to override the users' AM2R copy. Hence why we copy it additionally. Does that make sense?

IvarWithoutBones replied:
Makes sense :+1:

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+
+		# Format music for Unix, aka lowercase it
+		# Finds every file, and if its not already lowercase, lowercases it.
+		find "$RESOURCES" -type f -exec \
+			bash -c 'target="{}";
+				cd "$(dirname "${target}")";
+				target="$(basename "${target}")";
+				! [[ "${target}" = "${target,,}" ]] && mv "${target}" "${target,,}"' \;
```
I think this is a bit easier to read:
```suggestion
		find "$RESOURCES" -type f -exec bash -c '
		    target="{}"
		    cd "$(dirname "${target}")"
		    target="$(basename "${target}")"
		    ! [[ "${target}" = "${target,,}" ]] && mv "${target}" "${target,,}"
		' \;
```

Miepee replied:
Applied locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+
+		if [ "$SYSTEMWIDE" = true ]; then
+			mkdir "$PREFIX/bin"
+			ln -s "$OUTPUT/runner" "$PREFIX/bin/am2r"
```
Files should be copied to the `PREFIX` directory instead of being symlinked, as otherwise the `PREFIX` directory isn't independant. You should't need the `OUTPUT` directory after this script completes.

Miepee replied:
The prefix is currently completely independent though.
When doing a patch, this is the resulting file structure:
```
PREFIX
|-bin/
   |-am2r (symlink to PREFIX/opt/runner)
|-opt/
   |-assets/
   |-runner
|-share/
   |- etc.
```

This is probably confusion due to what you mentioned somewhere below, that for the AppImage, I temporarily change the meaning of "OUTPUT"

IvarWithoutBones replied:
Ah, that does make sense.

> This is probably confusion due to what you mentioned somewhere below, that for the AppImage, I temporarily change the meaning of "OUTPUT"

That's a bit confusing as you said, should be changed I think.

Miepee replied:
Yep, did try to change it afterwards in a commit.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+		fi
+		cp "$SCRIPT_DIR/DesktopTemplate" "$desktopPath"
+
+		# replace [REPLACE] with OUTPUT.Replace("/", "\/") in desktop file
```
Any reason the template doesn't have this in the first place?

Miepee replied:
Because it's not possible to know what "OUTPUT" is before running the script.
Comment is probably misleading a bit, as its more c#-code, but basically what it's supposed to say, is that all `/` in `OUTPUT` get replaced with `\/`, because otherwise there were some kind of issues with running the desktop file that I can't remember right now.

IvarWithoutBones replied:
> is that all `/` in OUTPUT get replaced with `\/`

That might be because you're using sed, which needs to have backslashes escaped. Desktop files themselves should not need that.

Miepee replied:
Ah yeah, was a sed thing my bad. Was because $OUTPUT contains a file path (i.e. `./foo/bar`), but when expanded to the full sed command, it'd be `sed -i "s/\[REPLACE\]/./foo/bar/"` which sed interpreted as "replace '[REPLACE]' with '.'".

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+		if [ "$APPIMAGE" = false ]; then
+			# For non-appimage, the desktop file should point to runner
+			sed -i "s/AM2R.AppImage/runner/" "$desktopPath"
+			echo ""
+			echo "The operation was completed successfully. See you next mission!"
```
Can be used later in the script for newlines as well.
```suggestion
			echo -n "\nThe operation was completed successfully. See you next mission!"
```

Miepee replied:
Hm, prefer how this looks cleaner / easier to follow visible for most operations.
In here for example:
(everything nice indented)
```bash
	echo "Select your patch type:"
	echo ""
	echo "1 - Linux"
	echo "2 - Android"
	echo ""
	echo "Awaiting input:"
```
vs
(initial line is indented correct, following lines break indent)
```bash
        echo -n "Select your patch type:

1 - Linux
2 - Android

Awaiting input:"
```
or even worse:
(single line)
```bash
        echo -n "Select your patch type:\n\n1 - Linux\n2 - Android\n\nAwaiting input:"
```

IvarWithoutBones replied:
In my opinion the single line option does look the cleanest, but that's pretty subjective.

Miepee replied:
Will leave this as-is then. IMO at least it helps with readability seeing the output exactly as it'll be shown when run, and I don't think any system this is targeted for would have problems due to multiple echo's.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+		mv "$OUTPUT" "$SCRIPT_DIR/AM2R.AppDir/bin"
+		ARCH=x86_64 "$SCRIPT_DIR/utilities/appimagetool-x86_64.AppImage" -n AM2R.AppDir 2> /dev/null
+
+		mkdir "$OUTPUT"
```
Reusing `OUTPUT` like this makes the script a bit confusing to read IMO.

Miepee replied:
Tried to make this a bit clearer.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+		java -jar "$apktoolPath" d -f "$SCRIPT_DIR/utilities/android/AM2RWrapper.apk"
+		cp -R "$SCRIPT_DIR/utilities/android/assets/" "$SCRIPT_DIR/AM2RWrapper/"
+		echo "Editing apktool.yml..."
+	    sed -i "s/doNotCompress:/doNotCompress:\n- ogg/" "$SCRIPT_DIR/AM2RWrapper/apktool.yml"
```
Indentation seems off

Miepee replied:
Applied locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+		fi
+
+		echo "Packaging APK..."
+		echo "If Java JDK 8 or newer is not present, this will fail!"
```
It'd be nice if this were to be checked instead.

Miepee replied:
Same with xdelta too then? Since those patching processes fail currently as well if it is not installed.

Miepee replied:
Done locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+	echo "Awaiting input:"
+	read -n1 input
+	echo ""
+	if [ $input = "1" ]; then
```
This isn't safe.
```suggestion
	if [[ "${input}" = "1" ]]; then
```

Miepee replied:
applied locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+	echo "[y/n]"
+	read -n1 input
+	echo ""
+	if [ $input = "y" ]; then
```
Same as above. Doing `${input,,}` will also make this case insensitive, which is a bit nicer.
```suggestion
	if [[ "${input,,}" = "y" ]]; then
```

Miepee replied:
Applied locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+	echo ""
+	if [ $input = "1" ]; then
+		OSCHOICE="linux"
+	elif [ $input = "2" ]; then
```
```suggestion
	elif [[ "${input}" = "2" ]]; then
```

Miepee replied:
ditto.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+		echo "[y/n]"
+		read -n1 input
+		echo ""
+		if [ $input = "y" ]; then
```
Needs expansion as well.

Miepee replied:
Ditto.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+
+	patch_am2r
+	result=$?
+	echo "Press any key to quit..."
```
Why doesn't it quit immediately after completing?

Miepee replied:
Most of the interactive CLIs I saw just waited for you to confirm it quitting. So that in case the user executed the script by double clicking, it doesnt just suddenly vanish.
(FWIW the default Mac terminal also does this automatically, probably cause they also realized that people dont just want their scripts to suddenly dissapear.)

IvarWithoutBones replied:
Ah, I didn't consider people launching this through a GUI. I suppose that makes sense, seeing how it isn't done in headless mode.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+		exit 1
+		;;
+	*)
+		POSITIONAL_ARGS+=("$1") # save positional arg
```
Shouldn't this also throw on an unknown option?

Miepee replied:
Ah yup, it should.
I combined it with the case from above into one.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+fi
+
+patch_am2r
+exit $?
```
This doesn't do anything, it will automatically use the exit code from the last command due to `set -e`.
```suggestion
```

Miepee replied:
Applied locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,302 @@
+
+# check if necessary variables are provided
+if [ -z "$OSCHOICE" ]; then
+	echo "Missing argument! The OS needs to be provided! For further info use --help."
```
It'd be nice if the error message suggested the flag you need to use to set this.

Miepee replied:
Applied locally.

IvarWithoutBones commented:
You should also use `local` for variables in functions which aren't used outside of it, as to avoid global variables as much as possible. Those should not be capitalized.

Miepee committed:
Implement review suggestions

- put main routine into seperate function
- use locals
- make some input access safe
- use tempdir for appimage creation
- check whether Java / xdelta is installed
- print errors to stderr
- Improve APK creation by copying less, and using tempDirs

Co-Authored-By: Ivv <41924494+ivarwithoutbones@users.noreply.github.com>

Miepee commented:
Implemented most of the review sugestions.

IvarWithoutBones commented on README.md:
```diff
@@ -13,7 +13,7 @@ Include = /etc/pacman.d/mirrorlist
 
 Then install the following dependencies:  
-`sudo pacman -S --needed python xdelta3 jre8-openjdk lib32-libpulse`  
+`sudo pacman -S --needed xdelta3 jre8-openjdk lib32-libpulse`
```
You're using `sed` as well, which might be preinstalled on some distros but it is not apart of coreutils. It should be documented as a dependency I think. On NixOS we use the `gnused` package.

`unzip` is also a required dependency.

Also, isn't the java stuff only needed for android builds?

Miepee replied:
Oh, thought both unzip and sed were part of coreutils
I put those into the dependencies, don't have the time right now to double check that the package names are 100% correct on the other distros. Will do so later (or when someone complains).

IvarWithoutBones commented on README.md:
```diff
@@ -25,27 +25,27 @@ sudo apt update && sudo apt install libc6:i386 libncurses5:i386 libstdc++6:i386
 
 ## Linux and Android patching process
-To patch your copy of (Windows) AM2R v1.1, place the `AM2R_11.zip` (case-sensitive) file in the same folder as `patcher.py`. After installing the required dependencies for the version you would like to patch to, execute `patcher.py` via `python patcher.py`.  
+To patch your copy of (Windows) AM2R v1.1, place the `AM2R_11.zip` (case-sensitive) file in the same folder as `patcher.sh`. After installing the required dependencies for the version you would like to patch to, execute `patcher.sh` via `./patcher.sh`.
```
```suggestion
To patch your copy of (Windows) AM2R v1.1, place the `AM2R_11.zip` (case-sensitive) file in the same folder as `patcher.sh`. After installing the required dependencies for the version you would like to patch to, execute the patcher script with `./patcher.sh`.
```

Miepee replied:
Done locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,330 @@
+		# Check whether Xdelta is installed
+		if [ ! command -v xdelta3 &> /dev/null ] ; then
+			>&2 echo "Xdelta is not installed! Please install 'xdelta3' from your local package manager!"
+			return 1
```
Just to communicated this is a fatal problem a bit clearer:
```suggestion
			exit 1
```

Miepee replied:
Done locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,330 @@
+		echo "Applying AM2R xdelta patch..."
+		xdelta3 -dfs "$OUTPUT/AM2R.exe" "$SCRIPT_DIR/data/AM2R.xdelta" "$OUTPUT/runner"
+
+		echo "Applying data xdelta patch..."
```
Im assuming this is talking about assets? If yes:
```suggestion
		echo "Applying asset xdelta patch..."
```
`data` is a bit unclear IMO, sounds a lot like the previous log message.

Miepee replied:
"data" in this case refers to the "data.win" file, which technically is *the* asset file.
Applied for less confusion.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,330 @@
+			cp "$SCRIPT_DIR"/data/HDR_HQ_in-game_music/*.ogg "$RESOURCES/"
+		fi
+
+		# Format music for Unix, aka lowercase it
+		# Finds every file, and if its not already lowercase, lowercases it.
+		find "$RESOURCES" -type f -exec bash -c '
+		    target="{}";
+		    cd "$(dirname "${target}")";
+		    target="$(basename "${target}")";
+		    ! [[ "${target}" = "${target,,}" ]] && mv "${target}" "${target,,}"
```
This is because of my previous CR comment being wrong, but the semicolons arent needed:
```suggestion
                # On Unix the music filenames need to be converted to lowercase.
		find "$RESOURCES" -type f -exec bash -c '
		    target="{}"
		    # Only files are meant to be modified, not the folders they're contained in.
		    cd "$(dirname "${target}")"
		    target="$(basename "${target}")"
		    # Convert the filename to lowercase, if required.
		    ! [[ "${target}" = "${target,,}" ]] && mv "${target}" "${target,,}"
```
Also breaks up the comment a bit

Miepee replied:
Brought in with some slight grammar change so that Kate doesn't complain.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,330 @@
+		sed -i "s/\[REPLACE\]/${OUTPUT//\//\\\/}/" "$desktopPath"
+
+		# If we don't want to create an AppImage, we're done.
+		if [ "$APPIMAGE" = false ]; then
```
The message saying the operation was succesfull could be moved to the bottom, so that it also gets printed when creating an appimage. Something like:
```suggestion
		if [ "$APPIMAGE" = true ]; then
                    ...
                else
                    sed -i "s/AM2R.AppImage/runner/" "$desktopPath"
                fi
                
               	echo "The operation was completed successfully. See you next mission!"
 ```

Miepee replied:
Didn't think of that. Applied locally.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,330 @@
+
+		# Create AppImage
+		echo "Creating AppImage..."
+		local tempAppDir=$(mktemp -d -u)
```
Why is `-u` used? From the manpage:
```man
       -u, --dry-run
              do not create anything; merely print a name (unsafe)
```

Also, when creating a temporary directory you should always trap its deletion. Currently, if any command before the `rm -R "$tempAppDir` fails it wont get deleted. You can automatically run commands upon exit with
```bash
trap "rm -rf $tempAppDir" EXIT
```

Miepee replied:
Because if the folder exists, then the copy operation from right below will copy *into* the folder, rather than move *as* the folder.
Added a comment to make it clearer.

IvarWithoutBones commented on patcher.sh:
```diff
@@ -0,0 +1,330 @@
+
+	# check if necessary variables are provided
+	if [ -z "$OSCHOICE" ]; then
+		>&2 echo "Missing argument! The OS needs to be provided via the \"--os\" flag! For further info use --help."
```
We've already discussed this privately, but I think making the default Linux makes a lot of sense. Most users will want to use that.

Miepee replied:
Done locally.

Miepee committed:
Applied more review comments

- mention the new dependencies
- make linux default OS
- make patcher not create in a subdir into prefix with local
- change return to exit for non-0
- better indicate arguments
- remove unneded dirs from cleaning

Co-Authored-By: Ivv <41924494+ivarwithoutbones@users.noreply.github.com>

Miepee committed:
Fixes mostly android bugs

- Removed credits
- make android builds not have unnecessary assets
- fix "is-installed" check
- Make android properly ignore systemwide

Miepee committed:
Some more slight cleanups

- rename OUTPUT to GAMEDIR to be more understandable

Miepee committed:
Fix application icon not being according to XDG in systemwide install

IvarWithoutBones reviewed (approved):
Looks good to me, thanks for the work! This really helps out with building AM2R using Nix, I plan to add the package to the nixpkgs repository as soon as this gets merged 👍

Lojemiru commented on patcher.sh:
```diff
@@ -0,0 +1,339 @@
+		GAMEDIR="${PREFIX}/opt/am2r"
+		RESOURCES="${GAMEDIR}/assets"
+	else
+		GAMEDIR="${PREFIX}"
```
This line is, to my understanding, substantially dangerous. $GAMEDIR gets rm -rf'd on line 43, and any directory name can be passed into this script as $PREFIX as per line 302. Say I'm a bit of a dimwitted user who works out of the desktop and decides to pass it in as the prefix - whole thing gets wiped just to dedicate the entire space to AM2R lol.

Miepee replied:
Addressed in latest commit.

Miepee committed:
Make accidental deletion of important folders impossible
